File: version_2/models/fsa_high_res/_phi_burst.py
"""Sub-daily Φ-burst expansion — JAX-vectorised, deterministic.

Converts a per-day Φ schedule (`(n_days,)` array of daily training-strain
rates) to a per-15-min Φ(t) schedule (`(n_days·96,)` array) with a
**morning-loaded** Gamma-shape activity profile per wake-window:

    t = hour_of_day − wake_hour      (zero at wake)
    shape(t) = t · exp(−t / τ)        (Gamma(k=2) shape)
                                       peaks at t = τ (~3h post-wake)

Normalised so each day's integrated Φ equals 24 · Φ_daily — i.e. the
slow Banister dynamics see the same daily load whether Φ is constant
or burst-shaped.

Sleep hours [sleep_hour, wake_hour+24]: Φ = 0.

Differences from public-dev `simulation.py:generate_phi_sub_daily`:
- JAX `vmap` over days × precomputed per-bin shape (no Python loop).
- **No multiplicative noise** (the public-dev default adds 15%); here Φ
  is a deterministic control output, no realism noise needed.

Used by:
- `simulation.py:generate_phi_sub_daily` (numpy reference, kept for
  parity with public-dev callers)
- `_plant.py:StepwisePlant.advance` (closed-loop control hook)
- `tools/bench_smc_*_fsa.py` drivers (open-loop and closed-loop)
"""

from __future__ import annotations

import jax
import jax.numpy as jnp
import numpy as np


# Time-grid constants (must match simulation.py)
DT_BIN_DAYS  = 1.0 / 96.0
DT_BIN_HOURS = 24.0 / 96.0
BINS_PER_DAY = 96


def _build_per_day_envelope(wake_hour: float = 7.0,
                              sleep_hour: float = 23.0,
                              tau_hours: float = 3.0) -> jnp.ndarray:
    """Construct the per-day Φ envelope (shape (96,) in JAX, sums to 24).

    Returns an array `e[k]` for k ∈ [0, 96) such that for any daily-Φ
    value `Φ_d`, the per-bin Φ(t_k) = `Φ_d · e[k]`.
    Specifically the envelope is normalised so `sum(e[k] * dt_hours) = 24`
    regardless of the wake/sleep window — preserving the daily integral.
    """
    h = jnp.arange(BINS_PER_DAY, dtype=jnp.float64) * DT_BIN_HOURS
    in_wake = (h >= wake_hour) & (h < sleep_hour)
    t_post = jnp.where(in_wake, h - wake_hour, 0.0)
    raw_shape = jnp.where(in_wake, t_post * jnp.exp(-t_post / tau_hours), 0.0)
    # Normalise so daily integral = 24 · 1 = 24
    daily_integral = (raw_shape * DT_BIN_HOURS).sum()
    envelope = raw_shape * (24.0 / jnp.maximum(daily_integral, 1e-12))
    return envelope


@jax.jit
def expand_daily_phi_to_subdaily_jax(daily_phi: jnp.ndarray,
                                       wake_hour: float = 7.0,
                                       sleep_hour: float = 23.0,
                                       tau_hours: float = 3.0) -> jnp.ndarray:
    """JAX-vectorised expansion: `(n_days,) → (n_days·96,)`.

    Each daily Φ value is multiplied by the precomputed per-day envelope.
    Result is a flat per-15-min array compatible with the v2 simulator's
    `aux = (Phi_arr,)` contract.
    """
    envelope = _build_per_day_envelope(wake_hour, sleep_hour, tau_hours)
    # Broadcast: (n_days, 1) × (1, 96) → (n_days, 96) → flatten
    out = jnp.outer(daily_phi, envelope)
    return out.reshape(-1)


def expand_daily_phi_to_subdaily(daily_phi: np.ndarray,
                                   wake_hour: float = 7.0,
                                   sleep_hour: float = 23.0,
                                   tau_hours: float = 3.0) -> np.ndarray:
    """Numpy convenience wrapper around the JAX implementation."""
    out_jax = expand_daily_phi_to_subdaily_jax(
        jnp.asarray(daily_phi, dtype=jnp.float64),
        wake_hour=wake_hour, sleep_hour=sleep_hour, tau_hours=tau_hours,
    )
    return np.asarray(out_jax, dtype=np.float32)

File: version_3_julia/src/FSAv5/PhiBurst.jl
# FSAv5/PhiBurst.jl — sub-daily Φ-burst expansion.
#
# Maps line-by-line to:
#   • Python: `models/fsa_high_res/_phi_burst.py:46-92`
# (No LEAN counterpart — this is a deterministic envelope helper, not
# part of the SDE math; the LEAN reference covers the dynamics layer.)
#
# Per-day Φ envelope: morning-loaded Gamma(k=2) shape, peaks ~3h
# post-wake, normalised so the daily integrated Φ is preserved
# regardless of wake/sleep window.
#
# Used by:
#   - `Plant.jl:advance!` (closed-loop control hook)
#   - the bench drivers when expanding daily-rate decisions to per-bin
#     stimulus arrays for the simulator.

# ── Time-grid constants (15-min bins, FSA_STEP_MINUTES=15 default) ─────────

const STEP_MIN_DEFAULT = 15

const BINS_PER_DAY = (60 * 24) ÷ STEP_MIN_DEFAULT          # 96
const DT_BIN_DAYS  = 1.0 / BINS_PER_DAY                     # 1/96
const DT_BIN_HOURS = 24.0 / BINS_PER_DAY                    # 0.25 h

# ── Per-day Φ envelope ─────────────────────────────────────────────────────

"""
    build_per_day_envelope(; wake_hour=7.0, sleep_hour=23.0, tau_hours=3.0)::Vector{Float64}

Construct the per-day Φ envelope of length `BINS_PER_DAY` (96 by default).

Mirrors `_phi_burst._build_per_day_envelope`. Returns array `e[k]`
such that for any daily-Φ value `Φ_d`, the per-bin Φ(t_k) = `Φ_d ·
e[k]`. The envelope is normalised so the daily integral
`sum(e[k] * dt_hours) = 24`, preserving the slow-Banister daily load.
"""
function build_per_day_envelope(; wake_hour::Float64  = 7.0,
                                  sleep_hour::Float64 = 23.0,
                                  tau_hours::Float64  = 3.0)::Vector{Float64}
    h = collect((0:BINS_PER_DAY-1) .* DT_BIN_HOURS)
    out = zeros(Float64, BINS_PER_DAY)
    @inbounds for k in 1:BINS_PER_DAY
        h_k = h[k]
        if h_k >= wake_hour && h_k < sleep_hour
            t_post = h_k - wake_hour
            out[k] = t_post * exp(-t_post / tau_hours)
        end
    end
    daily_integral = sum(out) * DT_BIN_HOURS
    if daily_integral > 1e-12
        out .*= (24.0 / daily_integral)
    end
    return out
end

# ── Daily → sub-daily expansion ────────────────────────────────────────────

"""
    expand_daily_phi_to_subdaily(daily_phi::AbstractVector{Float64};
                                  wake_hour=7.0, sleep_hour=23.0,
                                  tau_hours=3.0)::Vector{Float64}

Expand a per-day Φ schedule of length `n_days` into a per-bin Φ(t)
schedule of length `n_days * BINS_PER_DAY`. Each daily Φ value is
multiplied by the per-day envelope.

Mirrors `_phi_burst.expand_daily_phi_to_subdaily`.
"""
function expand_daily_phi_to_subdaily(daily_phi::AbstractVector{Float64};
                                       wake_hour::Float64  = 7.0,
                                       sleep_hour::Float64 = 23.0,
                                       tau_hours::Float64  = 3.0)::Vector{Float64}
    envelope = build_per_day_envelope(; wake_hour, sleep_hour, tau_hours)
    n_days   = length(daily_phi)
    out      = Vector{Float64}(undef, n_days * BINS_PER_DAY)
    @inbounds for d in 1:n_days
        phi_d = daily_phi[d]
        base  = (d - 1) * BINS_PER_DAY
        for k in 1:BINS_PER_DAY
            out[base + k] = phi_d * envelope[k]
        end
    end
    return out
end

export BINS_PER_DAY, DT_BIN_DAYS, DT_BIN_HOURS
export build_per_day_envelope, expand_daily_phi_to_subdaily
